### ISSUE TITLE
Global menu

### ISSUE THREAD

**Issue by Alexays** [Alexays/Waybar#288]

**Comment by varac** [comment]
What is this about ?

**Comment by Alexays** [comment]
Look at kde global menu

**Comment by ripefig** [comment]
Any progress on this? I'd recommend maybe not a global menu, but a window tile widget that when (right?) clicked brings up a popup palette with a list of all menu items. This way could even access menus for multiple recent windows, without having to focus each window before accessing the global menu. So you just have a window title widget and if the app exports menus there could be some indicator of that fact in the widget.

Also many programs don't use menus at all, so a classic "global menu" wouldn't really be global.

**Comment by Kommynct** [comment]
I feel like it'd make more sense for something like wofi to handle this.

**Comment by WillPower3309** [comment]
That would also be something really cool to see, and something I'd likely use on sway. That being said, a traditional global menu would be a really "nice to have" feature for any floating compositors that come out, and would allow waybar to be more compositor agnostic

**Comment by WillPower3309** [comment]
> Any progress on this? I'd recommend maybe not a global menu, but a window tile widget that when (right?) clicked brings up a popup palette with a list of all menu items. This way could even access menus for multiple recent windows, without having to focus each window before accessing the global menu. So you just have a window title widget and if the app exports menus there could be some indicator of that fact in the widget.
> 
> Also many programs don't use menus at all, so a classic "global menu" wouldn't really be global.
> 

That would also be something really cool to see, and something I'd likely use on sway. That being said, a traditional global menu would be a really "nice to have" feature for any floating compositors that come out, and would allow waybar to be more compositor agnostic

**Comment by puresoul** [comment]
Hi, got somehow same idea. Chose "nwg-menu" as base for my menu, but now have a little struggle with on which output will menu open. Because don't using focus with mouse in sway, menu open on workspace with focus and that no may be always the right one :-D

For now I hacked it with ugly double configuration for every output and running menu with right output as 
"static string" argument.

Is there some way how to dynamically (variable?) tell custom module on what output was executed?

Thanks

my hacked menu
https://github.com/puresoul/nwg-menu

**Comment by OJFord** [comment]
> What is this about ?

> Look at kde global menu

It seems it's a macOS-style menu bar (of File, View, etc.) for the whole workspace, replacing a per-window (on the window) bar, so you only see one - for the focussed window - and it's not necessarily anywhere near the window:

![KDE global menu](https://1.bp.blogspot.com/-YswFpE-Aoc0/XJ2cTaGH1UI/AAAAAAAAVU0/xJaUp015iTwyd5QtA4wLH6EZKwub6EzbgCLcBGAs/s1600/Screenshot_20190329_040630.png)

**Comment by henriquepicanco** [comment]
This would be a very, very, very (I've already said "very"?) awesome module for the Waybar.

**Comment by dmacvicar** [comment]
@Alexays I gave this a shot as I have been using labwc with Waybar, and one of the few things I am missing are global menus.

tl;dr: the whole appmenu story is kind of broken in Wayland if you are not on KDE / GNOME.

Getting a quick and dirty menu module to display the menu of an application was quite straightforward, as long as I hard-code the dbus menu service and path for a running application. All the required infrastructure is part of `GIO DbusMenuModel`.

Eg. for `transmission-gtk`

![grim-2023-11-18-183452](https://github.com/Alexays/Waybar/assets/15332/7cf034d4-cd03-4c07-8d26-6f5b2a622642)

The non-trivial part is for the module to get the current active window and from there the service/object path of the dbusmenu.

The original `com.canonical.AppMenu.Registrar` worked in a way apps would register those dbus properties with their Windows Id, so that the bar could keep the association. This is even unnecessary as the information is also set as xprops (`_KDE_NET_WM_APPMENU_OBJECT_PATH` / `_KDE_NET_WM_APPMENU_SERVICE_NAME` and `_GTK_APP_MENU_OBJECT_PATH` and `_GTK_MENUBAR_OBJECT_PATH`)

This does not work in Wayland. Apps would set bogus WindowId=1 or skip the registrar, and there are no xprops.

For Wayland, Plasma sets those props in a Qt extended surface protocol, but then the menu bar needs to talk to the compositor off-band to get this information, as there is no way to get this information from Waybar.

Gtk apps do something similar with another specific protocol (gtk-shell), which can't be used by Waybar. One would need the compositor to implement it and then Waybar would need a way to access this (IPC, like sway provides). This would tie the module with the compositor.

From there I suspect there are a few additional roadblocks to get the apps to automatically export their menus. Eg. keepassxc would only export them (under Wayland) when `com.canonical.AppMenu.Registrar` service is running (?). But I don't have the full picture here.

I started the module as a fork of the taskbar, hoping that the foreign top-level protocol would give me the starting point of activated windows. But the information it provides about the top-level is not enough to from there get to the dbusmenu services from that top-level.

I am a bit out of ideas 🤔. Perhaps I am missing something on how the protocols can help here.

Last thing floating in my mind was some kind of heuristic. Going from top-level app-id (provided by the protocol), to scan the whole session bus for the dbusmenu interface and try to match 🙈.

Some resources:

- https://docs.gtk.org/gio/class.DBusMenuModel.html
- https://blog.broulik.de/2016/10/global-menus-returning/
- https://wayland.app/protocols/kde-appmenu
- https://wayland.app/protocols/gtk-shell
- https://chromium-review.googlesource.com/c/chromium/src/+/3481665
- https://gitlab.com/vala-panel-project/vala-panel-appmenu/-/blob/master/subprojects/appmenu-gtk-module/src/platform.c?ref_type=heads

**Comment by probonopd** [comment]
Do you fully understand https://blog.broulik.de/2016/10/global-menus-returning/? (I don't.) It specifically talks about Wayland and I can confirm that the global menu implementation in KDE Plasma works under Wayland. Maybe it'd be helpful to run a KDE Plasma session, configure a panel with the global menu there, and use it for comparing what they are doing there?

**Comment by dmacvicar** [comment]
This is, as I understand it because  the extended surface protocol is between the app and the compositor (Kwin). The menu aplet can talk off band with Kwin (probably dbus).

An obvious path is the `foreign_toplevel` protocol. If you as a bar/taskbar implement this protocol, the compositor will tell you about the surfaces that become focused/toplevel and give you an API to activate them. This is eg. used by Waybar taskbar. It does not however provide the information to get the menu (DBUS service object path).

- A potential "right" fix, as I understand it, could be to implement both the Qt surface protocol and the Gtk shell protocol in wlroots, and extend the foreign toplevel protocol to carry this information to bars/taskbars/menubars. Then this would work with sway, labwc and others. I could be totally wrong here and would need to get some expert opinion.
- Another solution would be to hook into the toolkits and carry both the dbus info and the surface activation info off-band (eg. via dbus).

**Comment by probonopd** [comment]
I hope this can be done in a compatible way so that we don't end up with different protocols for achieving the same end result.

**Comment by dmacvicar** [comment]
I tried the offband approach too: built a gtk module that knows when any windows is focused. Plan was to extend the registrar to send this info, in addition to the dbus information.

It was going well until I ported to gtk-4 and horror :scream: : there [are no longer general-purpose gtk modules](https://gitlab.gnome.org/GNOME/gtk/-/issues/1132#note_300393) in gtk-4.

Following the discussion, the removal of the modules and hacks seems reasonable, but this [question](https://gitlab.gnome.org/GNOME/gtk/-/issues/1132#note_300721) was not addressed:
>What we lost and we are asking is how we can recover the possibility to auto-export the menu without depend of the decision of the developer of the Gtk application. How we can do by our self for example:

Conclusion:
- Gtk4 apps have a way to export the menu bar by using `GMenu`,which not all apps use (port from GtkMenu)
- Gtk4 apps do not seem to export them by default. Seems like an opt-in? (see  `gtk_application_impl_dbus_publish_menu` )

Can we even force a (ported to GMenu) app Gtk4 app to export its menus?

At this point I even forgot how I got the transmission app working with Waybar and a GMenu Dbus client (above), given that is linked to Gtk4.

**Comment by skela** [comment]
Might be some useful stuff in Fildem that might help here as well?
https://github.com/Weather-OS/Fildem-v2

I've got a global menu in qtile, and frankly i don't understand how people can use tiling window managers without them.
I very much want to use hyprland with the waybar, but the lack of a global menu feels wrong to me. 

Having a widget in the waybar that appears if the focused window has a detected menu, and a keyboard combo shortcut to bring up the menu. That would be fantastic.

**Comment by OJFord** [comment]
> I've got a global menu in qtile, and frankly i don't understand how people can use tiling window managers without them.

What's the connection? If anything I would think using a tiling wm makes it more likely the window is further away from the menu location, which is what I don't like about them even without tiling (e.g. using someone's Mac). Happily using sway without one, and i3 before that.

**Comment by dmacvicar** [comment]
Please lets keep the conversation focused. I am not collecting personal desktop preferences, but concrete information on how I can move forward the protoype I have for Waybar.

I posted a summary of what the challenges are:

* Waybar is not a compositor, therefore it can only get info about activated windows via [foreign_toplevel](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/protocol/wlr-foreign-toplevel-management-unstable-v1.xml?ref_type=heads) protocol, which only gives a handle and events on activation, but not info about their dbus interfaces. Some bars like KDE/Gnome extensions can get windows and dbus interface activity via the compositor because they have special APIs for it. Eg. TaskManager library in Plasma, wich Plasma appmenu applet uses.
* gtk3 apps need a special module, which does not exist (nor can exist) for gtk4. The module works with a service registering X11 dependent IDs
* gtk4 apps seem to expose the dbus information via a special gtk_shell protocol, which gnome-shell understands and has access to (being a compositor)
   * this menu is however, not exported by all applications automatically, and can't be influenced from the environment whether they do or not
   * when they do, it is the GIO DBusMenu protocol, which ironically, it is the easy to implement for waybar as a client.
* Qt apps are in good shape, because they export the menu, but in order to know the service location, you either use the X11-based registrar service, or you are compositor speaking the Qt surface protocol, which KDE does.

The best information I coult get is to dissect whether the following use-cases are possible, or whether someone has implemented them (all on pure-Wayland):

- a Qt app displaying a menu in some bar, not under KDE
- a Gtk4 app exporting the menu, not under Gnome
- a Gtk3 app, exporting the menu, not using the appmodule, or the menu not being displayed by a compositor plugin or extension, but a completely separate program

Otherwise, it would be best to start with labwc or an interface at the compositor level, which then bars can hook into (something like sway IPC).

**Comment by alexhulbert** [comment]
I think this should be possible on the hyprland wayland compositor with a hyprland plugin. Another plugin called "hyprriver" adds support for an extra wayland protocol and keeps track of metadata at the output level. The file that actually handles this is [here](https://github.com/zakk4223/hyprRiver/blob/main/RiverLayoutProtocolManager.cpp). Similar code could be written for [the qt surface extension protocol](https://github.com/KDE/plasma-wayland-protocols/blob/master/src/protocols/surface-extension.xml), which would allow the plugin to communicate the dbus object path and service name for the focused window to waybar. It would also allow [plasma-hud](https://github.com/Zren/plasma-hud) to work on wayland as well. I _almost_ want a global menu enough to write the plugin myself, but writing a hyprland plugin to work with the qt surface extension would be a little too much work for the effort. However, if someone else gets a working proof of concept to println the APPMENU_SERVICE_NAME and APPMENU_OBJECT_PATH of a given window when it makes a wayland protocol request, I'm happy to take it the rest of the way by setting up the process communication, waybar module, and getting something like plasma-hud to work. Also, I think some combination of the following allows non-qt windows to have the same KDE_WM appmenu properties added to them:
 - setting the `GTK_MODULES` env var to `appmenu-gtk-module` (with the associated package installed)
 - running `kded5` in the background and then loading the appmenu module with `qdbus org.kde.kded5 /kded org.kde.kded5.loadModule appmenu`
 - running the `gmenudbusmenuproxy` daemon

**Comment by Sodivad** [comment]
Hi we actually do not use qt-surface-extension for this (anymore? this was before I joined devleopment) but have our own protocol for [appmenu](https://invent.kde.org/libraries/plasma-wayland-protocols/-/blob/master/src/protocols/appmenu.xml?ref_type=heads) .
Related is the standardising effort https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52

**Comment by alexhulbert** [comment]
Oh wow so someone got this to work in something other than KDE! I got it _almost_ working last year and gave up after I couldn't figure out how to capture the `set_address` requests. I successfully captured the appmenu manager `create` requests, though. If anyone wants to take a crack at it, here's my code: https://github.com/alexhulbert/HyprAppMenu/blob/main/appMenuPlugin.cpp

I'm seeing messages like `APPMENU: created resource X for surface Y` when I open windows with appmenus, but I never see `APPMENU: set_address called`. @dkondor did you have to make a dbus service, set an env var, etc? Or maybe you can offer a debugging strategy?

Thanks!

**Comment by probonopd** [comment]
Thanks for this initiaitve. Global menus are the best (it has been proven scientifically).

Would it be possible to provide a working C++/Qt6, Python, or PyQt6 example on how to use this from an application? I'd like to use it in a desktop environment I am writing.

**Comment by alexhulbert** [comment]
> Thanks for this initiaitve. Global menus are the best (it has been proven scientifically).
> 
> Would it be possible to provide a working C++/Qt6, Python, or PyQt6 example on how to use this from an application? I'd like to use it in a desktop environment I am writing.

The linked code is just a Hyprland plugin designed to print the needed information to logs. Once I get the set_address call printing properly, I'm going to hook it up to a proper appmenu client.

I think this is what you're looking for: https://github.com/Zren/plasma-hud
This repo is a fairly minimal example of how you can use the dbus addresses from the wayland protocol to build an appmenu client.

**Comment by dkondor** [comment]
> Oh wow so someone got this to work in something other than KDE! I got it _almost_ working last year and gave up after I couldn't figure out how to capture the `set_address` requests. I successfully captured the appmenu manager `create` requests, though. If anyone wants to take a crack at it, here's my code: https://github.com/alexhulbert/HyprAppMenu/blob/main/appMenuPlugin.cpp
> 
> I'm seeing messages like `APPMENU: created resource X for surface Y` when I open windows with appmenus, but I never see `APPMENU: set_address called`. [@dkondor](https://github.com/dkondor) did you have to make a dbus service, set an env var, etc? Or maybe you can offer a debugging strategy?
> 
> Thanks!

My experience so far:
 - Note: I've only tested with GTK3 and Qt5 apps, so some of these might differ for the newer versions
 - For Qt apps, you need to set the `XDG_CURRENT_DESKTOP` environment variable to `KDE` (or something that starts with `KDE:`, it seems to be OK to have additional stuff there, but KDE seems to be required to be the first)
 - For GTK3 apps, you need to set the `GTK_MODULES` environment variable to include `appmenu-gtk-module` (and actually need to have this module installed :) )
 - You need to be running a "registrar" which is a DBus service that implements the `com.canonical.AppMenu.Registrar` interface and in theory allows apps that export menus to register themselves. AFAIK this is not really required (and not even usable) on Wayland, but Qt apps seem to only use `set_address` if such a service is running. You can find one implementation here: https://github.com/Cairo-Dock/cairo-dock-plug-ins/blob/master/Global-Menu/registrar/appmenu-registrar.py -- you only need this one Python script and can run it simply as `python3 appmenu-registrar.py`

**Comment by dkondor** [comment]
> Thanks for this initiaitve. Global menus are the best (it has been proven scientifically).
> 
> Would it be possible to provide a working C++/Qt6, Python, or PyQt6 example on how to use this from an application? I'd like to use it in a desktop environment I am writing.

I only have a GTK-based example so far: <br> https://github.com/dkondor/gtk_global_menu_test

This uses the protocol extensions I proposed (so will not work in general), but it has the code necessary to make menus work once you know the DBus addresses. In GTK, things are quite simple though, since the `org.gtk.Menus` interface is built-in to GTK and for the `com.canonical.dbusmenu` interface, you have the libdbusmenu-gtk3 library that takes care of everything.

**Comment by alexhulbert** [comment]
Thank you very much for the information, @dkondor ! I got it working! I should have some free time this weekend to make this into a plugin for Hyprland users. I'll start with a rofi-based command palette since that's a better fit for tiling window managers, but I imagine it should be easy enough to hook it up to a waybar module.

**Comment by bytefoundry-git** [comment]
Is it not possible to get the active window through AT-SPI (Assistive Technology Service Provider Interface)?

**Comment by M-Rick** [comment]
What I did is to launch XFCE4-Panel with Appmenu at startup. It is very stable. But I wanted to keep gnome Shell functions as well as well. At the beginning, I tried with Dash-to-Panel to reduce de panel size, but it was very buggy with the Dock that used to crash. In Ubuntu, the Dock is a modified version of Dash-to-Dock, and there is in incompatibility between them…
So, I created an extension that reduces se size of the panel. It works, well, but I can’t manage to align it well on the right. When a new button will be added to the Quick Settings, the button extends out of the screen instead of extending to the left… 
And I compiled the Gnome Shell Cosmic desktop tiling extension from PopOS!.
This solution works very well despite the little glitches, but unfortunately, it is X11 only…

https://github.com/M-Rick/MacUbuntu/

![Image](https://github.com/user-attachments/assets/182d62bd-2c9c-44e4-84d9-41e201000284)

**Comment by MachoXV** [comment]
You need to look at AxOS (another Arch Hyprland distro) menu and how they've done it. It's rather brilliant.

https://www.axos-project.com/

9 minutes into this video you have a great live example: https://youtu.be/j7uE4G8kcIQ?si=ahSpl1TzS0mxranu

**Comment by ilsubyeega** [comment]
> You need to look at AxOS (another Arch Hyprland distro) menu and how they've done it. It's rather brilliant.

@MachoXV Can you please put the screenshot of the global menu? I just skid the docs and didnt find it.

**Comment by alicealysia** [comment]
> You need to look at AxOS (another Arch Hyprland distro) menu and how they've done it. It's rather brilliant.
> 
> https://www.axos-project.com/
> 
> 9 minutes into this video you have a great live example: https://youtu.be/j7uE4G8kcIQ?si=ahSpl1TzS0mxranu

Watched through the whole video and didn't see an example of a global menu?

**Comment by ypurpl** [comment]
> You need to look at AxOS (another Arch Hyprland distro) menu and how they've done it. It's rather brilliant.
> 
> [axos-project.com](https://www.axos-project.com/)
> 
> 9 minutes into this video you have a great live example: [youtu.be/j7uE4G8kcIQ?si=ahSpl1TzS0mxranu](https://youtu.be/j7uE4G8kcIQ?si=ahSpl1TzS0mxranu)

this is not what this issue is talking about. getting it to show the name of the  currently focused application is already easy to do.

What this issue refers to is things like the menus in ~~Word~~ LibreOffice to show on your bar instead of in the app, displaying things like `File`, `Edit` `View` etc.

**Comment by M-Rick** [comment]
You can try LingmoOS as well. It seems to be KDE based.
https://lingmo.org/

**Comment by maxpoww** [comment]
Any news fellas, I been trying to workaround a global menu module for waybar but not look, I did not try so hard yet, now I'm in process of collecting information, so any glue you guys have will be really welcome! Thank in advance good people.

**Comment by probonopd** [comment]
Not related to Waybar, but I am implementing a global menu application using GNUstep at the moment: https://github.com/probonopd/gershwin-components/tree/main/Menu - maybe it is useful to someone.

**Comment by maxpoww** [comment]
Good people, anyone have something new on this matter?

**Comment by spydragon** [comment]
I too wish for news.
is it dead? still being worked on?
is there a plugin?
was the original post meant to signify it was a thing that was being worked on or as a suggestion that the author wanted to see if there was any interest in it?

**Comment by dkondor** [comment]
As far as I understand it, this needs explicit support on the protocol and in the compositor as well (note: there is already support in KWin for Qt apps, but not in other cases AFAIK). See also the discussion here: https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/merge_requests/131

In my opinion, an ideal way forward would be to first get the [foreign-toplevel-state](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/196) and [foreign-toplevel-management](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/156) finalized and merged (and start adopting in compositors, panels, etc.). After that my MR for wlr-protocols could be adapted. But I guess everyone has some more immediate stuff to work on, and the wlr-foreign-toplevel management protocol works already reasonably well (except for these kind of limitations of course).

**Comment by Sodivad** [comment]
Firefox also supports the kde protocol now https://github.com/mozilla-firefox/firefox/commit/1829d8679c7747d3decd3c3e6685c0a269a30b17 

You might need to toggle widget.gtk.global-menu.enabled and widget.gtk.global-menu.wayland.enabled

**Comment by dkondor** [comment]
Hi,
so I have an alternative implementation for Wayfire that uses its custom IPC instead of a new Wayland protocol:
https://github.com/WayfireWM/wayfire/pull/2959
Since there is no new protocol, this could provide an immediate solution (although limited to Wayfire only). See also a test client for how this could be made to work: https://github.com/dkondor/gtk_global_menu
